File: plugin/symboltest/testdata/goref/widget.go
// Package goref is a tiny, offline fixture module the Go reference symbol profile
// (plugin/symboltest.GoReferenceProfile) is driven against. It exercises the real
// declaration constructs the eight categories map to — a type, a function, a
// method, a constructor, a generic function, and a nested type — so
// goanalysis.IndexSymbols emits real symbols for them. It has no external
// dependencies and is loaded standalone (GOWORK=off).
package goref

// Widget is an exported struct type (category: types).
type Widget struct {
	name string
}

// Config is a type declared alongside Widget; the reference profile targets it as
// a Widget-nested declaration (category: nested declarations).
type Config struct {
	Verbose bool
}

// NewWidget is the Go NewT constructor idiom (category: constructors).
func NewWidget(name string) *Widget {
	return &Widget{name: name}
}

// Render is a method on *Widget (category: methods).
func (w *Widget) Render() string {
	return "<" + w.name + ">"
}

// GetName is an accessor-shaped method; the reference profile targets it as a
// generated symbol (category: generated symbols).
func (w *Widget) GetName() string {
	return w.name
}

// Build is a package-level function (category: functions).
func Build() *Widget {
	return NewWidget("default")
}

// Map is a generic function; the reference profile targets it with a
// type-parameter descriptor (category: overloads/generics).
func Map[T, U any](in []T, f func(T) U) []U {
	out := make([]U, 0, len(in))
	for _, v := range in {
		out = append(out, f(v))
	}
	return out
}

File: plugin/symboltest/testdata/dotnetref/Widget.cs
// Widget.cs is part of the offline dotnetref fixture: its declarations realize the
// §4.3 categories the .NET reference profile targets, so dotnetanalysis.IndexSymbols
// emits real symbols for them. It has no external dependencies and is read lexically
// (never built or run). Mirrors testdata/goref/widget.go for C#.
namespace Symboltest.DotNetRef
{
    // types (§4.3 row 2): the class type Widget.
    public class Widget
    {
        private readonly string _name;

        // constructors (§4.3 row 5): an explicit parameterless constructor; its
        // ECMA-335 metadata name is ".ctor".
        public Widget()
        {
            _name = "widget";
        }

        // methods (§4.3 row 4): an instance method taking one object parameter.
        public string Render(object value)
        {
            return _name + ":" + value;
        }

        // overloads/generics (§4.3 row 6): a 1-arity generic method whose identity is
        // the positional arity segment `1(string) — distinct from its non-generic
        // sibling below (which shares Name but differs in Descriptor).
        public T DeserializeObject<T>(string text)
        {
            return default;
        }

        // ...the non-generic sibling of the overload pair — same Name, no arity segment.
        public object DeserializeObject(string text)
        {
            return text;
        }

        // generated symbols (§4.3 row 8): an async method. The compiler synthesizes the
        // state-machine type <FetchAsync>d__0 (nested in Widget, Generated=true) — the
        // generated symbol the profile targets. It exists only in compiled IL, so the
        // pure-source scanner never observes it (SYMBOLS.md ⚑SM; PLAN-250 closes it).
        public async System.Threading.Tasks.Task<int> FetchAsync()
        {
            await System.Threading.Tasks.Task.Yield();
            return 0;
        }

        // nested declarations (§4.3 row 7): a type nested under Widget; its Enclosing is
        // the namespace + outer type.
        public class Inner
        {
            public void Ping()
            {
            }
        }
    }
}
